File: src/lib/software/hwmonitor.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class HWMonitor(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'hwmonitor'
        self.name = 'HWMonitor'
        self.category = SoftwareCategory.Utility
        self.download_name = 'hwmonitor.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = False
        self.icon = 'hwmonitor.png'
        self.homepage = 'https://cpuid.com/softwares/hwmonitor.html'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        html = reply.readAll().data().decode()

        version_pattern = compile(r'(\d\.\d+)\.exe')
        match = version_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(f'https://download.cpuid.com/hwmonitor/hwmonitor_{match.group(1)}.exe')

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self.homepage))

File: packages/app/src/services/software/definitions/hwmonitor.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'hwmonitor';
	public name = 'HWMonitor';
	public category = [SoftwareCategory.Utility];
	public downloadName = 'hwmonitor.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'hwmonitor.png';
	public homepage = 'https://cpuid.com/softwares/hwmonitor.html';

	public async resolveDownloadUrl() {
		const res = await fetch(this.homepage);
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/(\d\.\d+)\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(`https://download.cpuid.com/hwmonitor/hwmonitor_${match[1]}.exe`);
	}
}
